[LangRef] Rework DIExpression docs

slinder1 opened the pull request:
Factor out most of the DIExpression docs from LangRef.rst into SourceLevelDebugging.rst

What remains in LangRef is just enough context to make sense of how DIExpression-as-metadata fits into the IR, including some examples of the DIExpression syntax.

The rest now lives in the SourceLevelDebugging document, which gives more context to make sense of DIExpression-as-semantic-entity.

Use sections to clearly separate DWARF opcodes from LLVM internal-only opcodes, where before the distinction was only explicit in the source code. This was the original reason for the patch, and if the rest isn't accepted I can also just split up the list in LangRef instead.

Also make some other changes like fixing typos, using :ref: rather than unchecked links.

slinder1 committed:
[LangRef] Rework DIExpression docs

Factor out most of the DIExpression docs from LangRef.rst into
SourceLevelDebugging.rst

What remains in LangRef is just enough context to make sense of how
DIExpression-as-metadata fits into the IR, including some examples of
the DIExpression syntax.

The rest now lives in the SourceLevelDebugging document, which gives
more context to make sense of DIExpression-as-semantic-entity.

Use sections to clearly separate DWARF opcodes from LLVM internal-only
opcodes, where before the distinction was only explicit in the source
code.

Also make some other changes like fixing typos, using :ref: rather than
unchecked links, and rewording DW_OP_LLVM_entry_value.

t-tye commented on llvm/docs/SourceLevelDebugging.rst:
```diff
@@ -400,6 +404,189 @@ This intrinsic is equivalent to ``#dbg_assign``:
+  The next ``(N - 1)`` operations will be part of the ``DW_OP_entry_value``
+  block argument. For example, ``!DIExpression(DW_OP_LLVM_entry_value, 1,
+  DW_OP_plus_uconst, 123, DW_OP_stack_value)`` specifies an expression where
+  the entry value of ``reg`` is pushed onto the stack, and is added with 123.
```
The previous expression has no mention of ``reg``.

slinder1 replied:
The expression doesn't mention it, but the `DBG_VALUE` which uses the expression does (it is the single "location op", or in other words the "input" to the expression)

slinder1 replied:
I ended up redoing this almost entirely, as the original framing made any attempt to clarify things awkward.

One issue is that I don't know how to refer to "the implicit operation that all non-variadic debug expressions essentially start with", but in reality that is what the `DW_OP_LLVM_entry_value` operates on, and to be able to uniformly say it creates a sub-expression starting with the operation that precedes it seemed like the only way to really explain what `N` means.

Let me know what you think @OCHyams + @t-tye

t-tye commented on llvm/docs/SourceLevelDebugging.rst:
```diff
@@ -400,6 +404,189 @@ This intrinsic is equivalent to ``#dbg_assign``:
+  ``DW_OP_LLVM_arg, N`` refers to the ``N``\ :sup:`th` element in that list. For
+  example, ``!DIExpression(DW_OP_LLVM_arg, 0, DW_OP_LLVM_arg, 1, DW_OP_minus,
+  DW_OP_stack_value)`` used with the list ``(%reg1, %reg2)`` would evaluate to
+  ``%reg1 - reg2``. This list of values should be provided by the containing
```
evaluate to an implict value location that vas the value of ``%reg1 - reg2``.

t-tye commented on llvm/docs/SourceLevelDebugging.rst:
```diff
@@ -400,6 +404,189 @@ This intrinsic is equivalent to ``#dbg_assign``:
+  intrinsic/instruction.
+- ``DW_OP_LLVM_extract_bits_sext, 16, 8,`` specifies the offset and size
+  (``16`` and ``8`` here, respectively) of bits that are to be extracted and
+  sign-extended from the value at the top of the expression stack. If the top of
+  the expression stack is a memory location then these bits are extracted from
+  the value pointed to by that memory location. Maps into a ``DW_OP_shl``
```
What indicates if the top of the stack is a memory address or an integer value? Does this expand to an additional DW_OP_reref if it is a memory address?

OCHyams reviewed (approved):
Splitting the ops into sections makes sense and I think moving the details to SourceLevelDebugging is reasonable - there's precedence, as the debug record entry does the same.

LGTM, I'll let @t-tye Approve it when he's happy. I think most of the improvements @t-tye suggested apply to the existing text that you're just moving so it could be easier to review if that comes as a separate patch?

OCHyams commented:
> LGTM, I'll let @t-tye Approve it when he's happy.

Oops, I approved it accidentally despite what I said - please wait for an approval from @t-tye

t-tye commented on llvm/docs/SourceLevelDebugging.rst:
```diff
@@ -394,12 +398,231 @@ This intrinsic is equivalent to ``#dbg_assign``:
+  ``DW_OP_entry_value [reg], ...``, which pushes the value ``reg`` had upon
+  frame entry onto the DWARF expression stack.
+
+  Because ``DW_OP_LLVM_entry_value`` is currently limited registers, it is
```
limited registers -> limited to registers

OCHyams reviewed (commented):
(still LGTM)

slinder1 commented:
This conflicted with #153358 and #153932 when rebasing

I squashed and rebased, fixing up the relevant bits, and then extended the `inline-code`-ification from those patches through the entire `SourceLevelDebugging.rst` document. I added @kazutakahirata and @tgymnich in case they wanted to review. If I don't hear from them I'll probably merge early next week.

slinder1 commented:
> > I squashed and rebased, fixing up the relevant bits, and then extended the `inline-code`-ification from those patches through the entire `SourceLevelDebugging.rst` document. I added @kazutakahirata and @tgymnich in case they wanted to review. If I don't hear from them I'll probably merge early next week.
> 
> Thank you for looping me in and rebasing your PR. I don't have specific inputs, so don't let me block you. Thanks again!

Of course, and thank you for the quick response! (and for cleaning up LangRef)
